### PR TITLE
fourth set of modules

### PR DESCRIPTION
**Pull request by carchi8py**
##### SUMMARY
four set of 5
We are updating and adding the Netapp ONTAP specific modules. The update modules add new options for the existing functions. The new modules allow for setting up and modify different part of the ONTAP system.

We have about 33 modules, we've been told to start with 2 or 3 first, and take feedback on these first before pushing them all out

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
	na_ontap_lun.py
	na_ontap_net_ifgrp.py
	na_ontap_net_port.py
	na_ontap_net_routes.py
	na_ontap_net_vlan.py

##### ANSIBLE VERSION

```
ansible 2.6.0 (devel 0d1e9bfe90) last updated 2018/04/23 14:26:54 (GMT -400)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /tmp/ansible/lib/ansible
  executable location = /tmp/ansible/bin/ansible
  python version = 2.7.5 (default, Oct 11 2015, 17:47:16) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by schmots1** [comment]
shipit

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+
+  size_unit:
+    description:
+    - The unit used to interpret the size parameter.
+    choices: ['bytes', 'b', 'kb', 'mb', 'gb', 'tb', 'pb', 'eb', 'zb', 'yb']
+    default: 'gb'
+
+  force_resize:
+    description:
+    - Forcibly reduce the size. This is required for reducing the size of the LUN to avoid accidentally
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py]: multiline yaml string

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+    type: bool
+    default: false
+
+  vserver:
+    required: true
+    description:
+    - The name of the vserver to use.
+
+  ostype:
+    required: false
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    description:
+    - The os type for the LUN.
+    default: 'image'
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py]: `choices`?

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+
+  space_reserve:
+    required: false
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py]: redundant

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+            size=dict(type='int'),
+            size_unit=dict(default='gb',
+                           choices=['bytes', 'b', 'kb', 'mb', 'gb', 'tb',
+                                    'pb', 'eb', 'zb', 'yb'], type='str'),
+            force_resize=dict(default=False, type='bool'),
+            force_remove=dict(default=False, type='bool'),
+            force_remove_fenced=dict(default=False, type='bool'),
+            flexvol_name=dict(required=True, type='str'),
+            vserver=dict(required=True, type='str'),
+            ostype=dict(required=False, type='str', default='image'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py:169]: `choices`?

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 6]
```diff
@@ -0,0 +1,399 @@
...
+                return False
+            else:
+                self.module.fail_json(msg="Error resizing lun %s: %s" % (path, to_native(e)),
+                                      exception=traceback.format_exc())
+
+        return True
+
+    def apply(self):
+        property_changed = False
+        multiple_properties_changed = False
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py]: unused

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py [hunk 7]
```diff
@@ -0,0 +1,399 @@
...
+                    if not lun_exists:
+                        self.create_lun()
+
+                    else:
+                        if size_changed:
+                            # Ensure that size was actually changed. Please
+                            # read notes in 'resize_lun' function for details.
+                            size_changed = self.resize_lun()
+                            if not size_changed and not \
+                                    multiple_properties_changed:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_lun.py]: unused

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    - netapp.na_ontap
+version_added: '2.6'
+author:
+- Chris Archibald (carchi@netapp.com), Kevin Hutton (khutton@netapp.com), Suhas Bangalore Shekar (bsuhas@netapp.com)
+description:
+- Create, modify, destroy the network interface group
+options:
+  state:
+    description:
+    - Whether the specified volume should exist or not.
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py]: volume?

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+    description:
+    - Specifies the traffic distribution function for the ifgrp.
+    choices: ['mac', 'ip', 'sequential', 'port']
+
+  name:
+    description:
+    - Specifies the interface group name.
+    required: true
+
+  mode:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py:41]: `choices: [multimode, multimode_lacp, singlemode]`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+    - Specifies the link policy for the ifgrp.
+
+  node:
+    description:
+    - Specifies the name of node.
+    required: true
+
+  port:
+    description:
+    - Adds the specified port.
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py:52]: This module UI seems incorrect. It allows addition of one port at a time, but does not allow removal of a port at all (without trashing the ifgrp too). The "right" way to do this is to make this member a list- this is the same issue with people usually expecting to be able to work with both an absolute list, eg, "these and only these ports should be in the ifgrp" vs "ensure these ports are in the ifgrp but leave others not mentioned". Either is fine (possibly with a mode switch to support both), but the module's not terribly usable when you can only add one port at a time and not remove at all- it's confusing for the way people typically interact with Ansible modules.

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+    def __init__(self):
+        """
+            Initialize the Ontap IfGrp class
+        """
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present', 'absent'], default='present'),
+            distribution_function=dict(required=False, type='str', choices=['mac', 'ip', 'sequential', 'port']),
+            name=dict(required=True, type='str'),
+            mode=dict(required=False, type='str'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py:104]: `choices`

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+            node=dict(required=True, type='str'),
+            port=dict(required=False, type='str'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py:106]: should probably be a list as discussed above

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+short_description: Manage NetApp Ontap network ports.
+extends_documentation_fragment:
+    - netapp.na_ontap
+version_added: '2.6'
+author:
+- Chris Archibald (carchi@netapp.com), Kevin Hutton (khutton@netapp.com), Suhas Bangalore Shekar (bsuhas@netapp.com)
+description:
+- Modify a Ontap network port.
+options:
+  state:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:25]: Why does this arg exist, unless there's a future plan to support another state?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+    - Specifies the name of node.
+    required: true
+  port:
+    description:
+    - Specifies the name of port.
+    required: true
+  mtu:
+    description:
+    - Specifies the maximum transmission unit (MTU) reported by the port.
+  autonegotiate_admin:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:41]: `type: bool`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 15]
```diff
@@ -0,0 +1,233 @@
...
+        """
+            Initialize the Ontap Net Port Class
+        """
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present'], default='present'),
+            node=dict(required=True, type="str"),
+            port=dict(required=True, type="str"),
+            mtu=dict(required=False, type="str", default=None),
+            autonegotiate_admin=dict(required=False, type="str", default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:97]: `type: bool`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+    - Specifies the name of port.
+    required: true
+  mtu:
+    description:
+    - Specifies the maximum transmission unit (MTU) reported by the port.
+  autonegotiate_admin:
+    description:
+    - Enables or disables Ethernet auto-negotiation of speed,
+      duplex and flow control.
+  duplex_admin:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:45]: `choices: [auto, half, full]`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+    description:
+    - Specifies the user preferred duplex setting of the port.
+  speed_admin:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:48]: `choices: [auto, 10, 100, 1000, 10000]`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+    description:
+    - Specifies the user preferred speed setting of the port.
+  flowcontrol_admin:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:51]: `choices: [none, receive, send, full]`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 19]
```diff
@@ -0,0 +1,233 @@
...
+    def __init__(self):
+        """
+            Initialize the Ontap Net Port Class
+        """
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present'], default='present'),
+            node=dict(required=True, type="str"),
+            port=dict(required=True, type="str"),
+            mtu=dict(required=False, type="str", default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:96]: shouldn't this be `type='int'`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 20]
```diff
@@ -0,0 +1,233 @@
...
+            autonegotiate_admin=dict(required=False, type="str", default=None),
+            duplex_admin=dict(required=False, type="str", default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:98]: choices

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+            speed_admin=dict(required=False, type="str", default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:99]: choices

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 22]
```diff
@@ -0,0 +1,233 @@
...
+            flowcontrol_admin=dict(required=False, type="str", default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:100]: choices

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 23]
```diff
@@ -0,0 +1,233 @@
...
+        Modify a port
+        """
+        port_obj = netapp_utils.zapi.NaElement('net-port-modify')
+        port_obj.add_new_child("node", self.node)
+        port_obj.add_new_child("port", self.port)
+        # The following options are optional.
+        # We will only call them if they are not set to None
+        if self.mtu:
+            port_obj.add_new_child("mtu", self.mtu)
+        if self.autonegotiate_admin:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:177]: if you switch the argspec to `type: bool`, make sure to add `is not None` here

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py [hunk 24]
```diff
@@ -0,0 +1,233 @@
...
+        Run Module based on play book
+        """
+        changed = False
+        net_port_exists = False
+        results = netapp_utils.get_cserver(self.server)
+        cserver = netapp_utils.setup_na_ontap_zapi(
+            module=self.module, vserver=results)
+        netapp_utils.ems_log_event("na_ontap_net_port", cserver)
+        net_port_details = self.get_net_port()
+        if net_port_details:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_port.py:200]: shouldn't it be an error if the port *doesn't* exist? We're talking hardware here, right?

File: lib/ansible/modules/storage/netapp/na_ontap_net_routes.py [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+            Initialize the Ontap Net Route class
+        """
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=[
+                       'present', 'absent'], default='present'),
+            vserver=dict(required=True, type='str'),
+            destination=dict(required=True, type='str'),
+            gateway=dict(required=True, type='str'),
+            metric=dict(required=False, type='str', default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_routes.py:99]: `type='int'`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_routes.py [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+            new_destination=dict(required=False, type='str', default=None),
+            new_gateway=dict(required=False, type='str', default=None),
+            new_metric=dict(required=False, type='str', default=None),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_routes.py:102]: `type='int'`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_routes.py [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+                    changed = True
+        else:
+            if self.state == 'present':
+                changed = True
+
+        if changed:
+            if self.module.check_mode:
+                pass
+            else:
+                if not route_exists:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_routes.py:224]: The "modification" case is not idempotent. If you were to run the modify task twice, the first time would modify as expected, the second time would re-create the original route. If `new_X` is specified, it should assume it's running in a modification mode- if the original value doesn't exist, ensure that the `new_X` version does, then return `changed: false`, else error (invalid route or something).

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 28]
```diff
@@ -0,0 +1,198 @@
...
+extends_documentation_fragment:
+    - netapp.ontap
+version_added: '2.6'
+author: Chris Archibald (carchi@netapp.com), Kevin Hutton (khutton@netapp.com)
+description:
+- Create or Delete a network vlan
+options:
+  state:
+    description:
+    - Whether you want to create to delete a network vlan
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py]: Probably more like "Whether the specified network vlan should exist or not"

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 29]
```diff
@@ -0,0 +1,198 @@
...
+    choices: ['present', 'absent']
+    default: present
+  parent_interface:
+    description:
+    - The interface that hosts the vlan interface.
+    required: true
+  vlanid:
+    description:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py:33]: `type: int`?

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 30]
```diff
@@ -0,0 +1,198 @@
...
+    - The vlan id. Ranges from 1 to 4094.
+    required: true
+  node:
+    description:
+    - Node name of vlan interface.
+  interface_name:
+    description:
+    - Name of vlan interface. The name must be of the format <parent-inteface>-<vlanid>
+  gvrp_enabled:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py:42]: `type: bool`? `default: false` (and remove the redundant default text)

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 31]
```diff
@@ -0,0 +1,198 @@
...
+        Initializes the NetAppOntapVlan function
+        """
+        self.argument_spec = netapp_utils.ontap_sf_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present', 'absent'], default='present'),
+            parent_interface=dict(required=True, type='str'),
+            vlanid=dict(required=True, type='str'),
+            node=dict(required=True, type='str'),
+            interface_name=dict(required=False, type='str'),
+            gvrp_enabled=dict(required=False, type='str'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py]: `type='bool'`

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 32]
```diff
@@ -0,0 +1,198 @@
...
+    """
+    def __init__(self):
+        """
+        Initializes the NetAppOntapVlan function
+        """
+        self.argument_spec = netapp_utils.ontap_sf_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, choices=['present', 'absent'], default='present'),
+            parent_interface=dict(required=True, type='str'),
+            vlanid=dict(required=True, type='str'),
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py:82]: `type='int'`

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 33]
```diff
@@ -0,0 +1,198 @@
...
+        """
+        Deletes a vland
+        """
+        vlan_obj = netapp_utils.zapi.NaElement("net-vlan-delete")
+        vlan_info = self.create_vlan_info()
+
+        vlan_obj.add_child_elem(vlan_info)
+        self.server.invoke_successfully(vlan_obj, True)
+
+    def does_vlan_exits(self):
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py]: s/exits/exist/ ?

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py [hunk 34]
```diff
@@ -0,0 +1,198 @@
...
+        changed = False
+        result = None
+        results = netapp_utils.get_cserver(self.server)
+        cserver = netapp_utils.setup_ontap_zapi(module=self.module, vserver=results)
+        netapp_utils.ems_log_event("na_ontap_net_vlan", cserver)
+        existing_vlan = self.does_vlan_exits()
+        if existing_vlan:
+            if self.state == 'absent':  # delete
+                changed = True
+        else:
```
Comment (nitzmahone) [inline lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py]: should probably be validating whichever components aren't part of the vlan's "primary key" (seems like parent interface/node/interface_name, but not sure off the top of my head) and either allow them to be modified or at least warn/error if they're not in the requested state

**Review by nitzmahone — CHANGES_REQUESTED** [review]
a few minor changes in some, but the ifgrp module probably needs some overhaul

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py (hunk 1 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_lun.py]: Fixed

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py (hunk 2 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_lun.py]: Fixed

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py (hunk 6 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_lun.py]: Fixed.

File: lib/ansible/modules/storage/netapp/na_ontap_lun.py (hunk 7 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_lun.py]: Fixed.

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py (hunk 8 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py]: Fixed.

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py (hunk 29 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py:33]: Even though it a number our zapi treats it as a string and will fail if a int is passed.

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py (hunk 30 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py:42]: Fixed.

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py (hunk 31 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py]: Fixed

File: lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py (hunk 33 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_vlan.py]: Fixed.

File: lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py (hunk 10 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_ifgrp.py:52]: We will look in to this for the next release.

File: lib/ansible/modules/storage/netapp/na_ontap_net_port.py (hunk 24 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_port.py:200]: This is something we'll look at fixing in the new version.

File: lib/ansible/modules/storage/netapp/na_ontap_net_routes.py (hunk 27 above)
Comment (carchi8py) [reply lib/ansible/modules/storage/netapp/na_ontap_net_routes.py:224]: We'll look into fixing this in the next release.

**Pull request merged**